### ISSUE TITLE
click open popup button in datepicker second time, the popup is NOT 'year' mode anymore.

### ISSUE THREAD

**Issue by redguardtoo** [angular-ui/bootstrap#5978]
The issues forum is __NOT__ for support requests. It is for bugs and feature requests only.
Please read https://github.com/angular-ui/bootstrap/blob/master/CONTRIBUTING.md and search
existing issues (both open and closed) prior to opening any new issue and ensure you follow the instructions therein.

### Bug description:
set datepicker's datepicker-mode to 'year', `<input type="text" datepicker-mode"'year'" is-open="isOpen" /><button ng-click="isOpen = true">Open</button>`, when click the open button, the popup is RIGHTLY opened with `year` mode. Manually clean the date inside the input box, click the button again. This time the popup is displayed but NOT in `year` mode.

### Version of Angular, UIBS, and Bootstrap

Angular: 1.4.8

UIBS: 0.13.4 and 0.14.3

Bootstrap: 3.3.6

**Comment by wesleycho** [comment]
Not able to reproduce in Plunker [here](http://plnkr.co/edit/WwcTudNz9PqstqwfkG6Z?p=preview).

In the future, please do not ignore the section to add a reproduction in Plunker in the issue template - we provide convenient links in the documentation page for creating Plunkers with everything set up, and it would save us very significant amount of time chasing down potential issues.

**Comment by TechCraver1** [comment]
@wesleycho I do not get what you are saying...

**Comment by redguardtoo** [comment]
I can reproduce the issue in your plunkcer,
- in the first datepicker control, click the icon on left side, year popup is RIHGTLY displayed. Continue to finish the date input
- move focus to input box and manually clear the date
- press icon again, I expect to see the year popup, instead I got the WRONG date popup.

Firefox 46.0.1

**Comment by josetaira** [comment]
From the documentation:

> datepickerMode (Default: day) - Current mode of the datepicker (day|month|year). Can be used to initialize the datepicker in a specific mode.

Datepicker mode seems to retain the last mode that's used. So in your replication steps, if you "continue to finish the date input", you'd be at `day` mode. When you open again, it'd show the day mode view. If you want to force year mode again, you should change datepickerMode's value.

Also, setting ```datepicker-mode"'year'"``` doesn't seem to be supported. The value is updated in the options.

**Comment by redguardtoo** [comment]
Yes, as you can see, this issue can be reproduced. If I set to 'year' mode, it should not changed to 'day' mode if the the input box is empty.

This issue is reported by our client's QA engineer. When user input the date of birth on mobile, the best mode is 'year' mode when the date input box is empty because user only need press button 3 times.

**Comment by wesleycho** [comment]
This would involve us doing some sort of caching in order to retain what the setting should be - however, this attribute can be changed programmatically, so any cached value would be overridden by any updated value.

I guess we could update the cache every time a watcher fires (would need to add that code), then reset it to the previous cached value...but this logic is convoluted. I'm willing to consider a well-tested PR (would need tests to cover all possible scenarios), but not wholly convinced this is a good thing in the library itself.

**Comment by josetaira** [comment]
@wesleycho I created a PR https://github.com/angular-ui/bootstrap/pull/6022 that does this:
1. Store the initial value of $scope.datepickerMode (after it's set)
2. On $scope.$destroy, reset the values of $scope.datepickerMode and $scope.datepickerOptions.datepickerMode to the stored value.
